lib/micromachine.rb

```diff
@@ -44,9 +49,21 @@ def states
 private
 
   def change(event)
-    @state = transitions_for[event][@state]
+    previous_state, @state = @state, transitions_for[event][@state]
+
+    all_arguments = [event, previous_state, state]
+
     callbacks = @callbacks[@state] + @callbacks[:any]
-    callbacks.each { |callback| callback.call(event) }
+    callbacks.each do |callback|
+      case callback.arity
+      when -1
+        callback.call(*all_arguments)
+      when 0..3
+        arguments = all_arguments.take(callback.arity)
```

Comment:
Yeah; I’d done exactly that for #27 (increased branching). You’re guaranteeing at least `n+1` Array allocations where `n` is the number of callbacks. This is probably small, but depending on how the state machine is used…it may be important to ensure optimization here. And no, I don’t actually have a clue as to whether the branching or allocations are more expensive in performance; that’s what benchmarks are for.